File: Javascript/String/E-14. Longest Common Prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
  const prefix = strs[0].split('');
  for (const s of strs) {
    for (let i = 0; i < prefix.length; i++) {
      if (prefix[i] !== s[i]) prefix.splice(i);
      if (prefix.length - 1 < i) break;
    }
  }
  return prefix.join('');
};

const testData = [
  {
    strs: ['flower', 'flow', 'flight'],
    ans: 'fl',
  },
  {
    strs: ['dog', 'racecar', 'car'],
    ans: '',
  },
];

for (const data of testData) {
  console.log(data.ans === longestCommonPrefix(data.strs));
}

File: String/E-14. Longest Common Prefix/E-14. Longest Common Prefix.py
class Solution:
    @classmethod
    def longestCommonPrefix(self, strs):
        prefix = strs[0]

        for i in range(len(prefix)):
            for s in strs:
                if i > len(s) - 1 or prefix[i] != s[i]:
                    prefix = prefix[:i]
                if len(prefix) - 1 < i:
                    return prefix
        return prefix


test_data = [
    {
        "strs": ['flower', 'flow', 'flight'],
        "ans": 'fl',
    },
    {
        "strs": ['dog', 'racecar', 'car'],
        "ans": '',
    },
    {"strs": ["ab", "a"], "ans": "a"}
]

for data in test_data:
    print(data["ans"] == Solution.longestCommonPrefix(data["strs"]))
